File: endpoints/julia/reliable.jl
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 ZeroDDS Contributors
#
# Reliable XRCE stream (spec §8.4.10/§8.4.11) for the Julia endpoint SDK.
# Self-contained: the sender/receiver state machines + the HEARTBEAT / ACKNACK
# frame codec, byte-identical to `crates/xrce` and the other endpoint SDKs.
# Importable beside `zerodds.jl`; the reliable WRITE_DATA frame is the same
# 8-byte header as `ZeroDDS.write_frame(0x80, 0x80, seq, sample)`.
#
# Frame layout (little-endian): [0]=session [1]=stream [2..4]=seq LE
#   [4]=submessage id [5]=flags [6..8]=body-len LE [8..]=body
#   WRITE_DATA id=0x07 flags=0x03 body=sample; header seq = RFC-1982 sample seq
#   HEARTBEAT  id=0x0B flags=0x01 body(5)= first i16 LE, last i16 LE, stream
#   ACKNACK    id=0x0A flags=0x01 body(5)= first_unacked i16 LE, nack[0], nack[1], stream
#
# Wrapped in `module Reliable` to avoid clashing with `ZeroDDS`'s own names
# (`write_frame`, `Client`, ...).

module Reliable

export SESSION_NOKEY, STREAM_NONE, RELIABLE_STREAM_ID, HEARTBEAT_PERIOD_MS,
       SENDER_WINDOW, RECEIVER_BUFFER, MAX_PAYLOAD,
       SubmitStatus, SR_OK, SR_TOO_LARGE, SR_WINDOW_FULL,
       Sender, Receiver,
       seq_lt, seq_gt,
       reliable_write_frame, heartbeat_frame, acknack_frame,
       parse_heartbeat, parse_acknack,
       in_flight_count, submit!, get_in_flight, in_flight_pairs,
       pending_heartbeat!, recv_acknack!,
       expected, out_of_order_count, recv_data!, drain_in_order!, pending_acknack,
       reset!

const SESSION_NOKEY = 0x80
const STREAM_NONE = 0x00
const RELIABLE_STREAM_ID = 0x80
const HEARTBEAT_PERIOD_MS = 500
const SENDER_WINDOW = 16
const RECEIVER_BUFFER = 64
const MAX_PAYLOAD = 65535
const SM_WRITE_DATA = 0x07
const SM_ACKNACK = 0x0A
const SM_HEARTBEAT = 0x0B
const FLAG_WRITE = 0x03
const FLAG_E_LE = 0x01

# --- RFC-1982 16-bit serial-number comparison ---

"`a < b` in RFC-1982 order (half-window = 32768)."
function seq_lt(a::UInt16, b::UInt16)
    d = (UInt32(b) - UInt32(a)) & 0xffff
    d != 0x00000000 && d < 0x00008000
end

seq_gt(a::UInt16, b::UInt16) = seq_lt(b, a)

# --- frame codec ---

"Reliable WRITE_DATA — identical bytes to `ZeroDDS.write_frame(0x80, 0x80, seq, sample)`."
function reliable_write_frame(seq::UInt16, sample::Vector{UInt8})
    n = length(sample)
    hdr = UInt8[
        SESSION_NOKEY, RELIABLE_STREAM_ID,
        UInt8(seq & 0xff), UInt8((seq >> 8) & 0xff),
        SM_WRITE_DATA, FLAG_WRITE,
        UInt8(n & 0xff), UInt8((n >> 8) & 0xff),
    ]
    vcat(hdr, sample)
end

"byte-identical to `golden_heartbeat_le.bin` for (0x80, 0x00, 1, {1,3,0x80})."
function heartbeat_frame(session::UInt8, stream_hdr::UInt8, msg_seq::Integer,
                          first::UInt16, last::UInt16, stream_id::UInt8)
    UInt8[
        session, stream_hdr,
        UInt8(msg_seq & 0xff), UInt8((msg_seq >> 8) & 0xff),
        SM_HEARTBEAT, FLAG_E_LE, 0x05, 0x00,
        UInt8(first & 0xff), UInt8((first >> 8) & 0xff),
        UInt8(last & 0xff), UInt8((last >> 8) & 0xff),
        stream_id,
    ]
end

"byte-identical to `golden_acknack_le.bin` for (0x80, 0x00, 1, {1,0,0,0x80})."
function acknack_frame(session::UInt8, stream_hdr::UInt8, msg_seq::Integer,
                        first_unacked::UInt16, nack_lo::UInt8, nack_hi::UInt8, stream_id::UInt8)
    UInt8[
        session, stream_hdr,
        UInt8(msg_seq & 0xff), UInt8((msg_seq >> 8) & 0xff),
        SM_ACKNACK, FLAG_E_LE, 0x05, 0x00,
        UInt8(first_unacked & 0xff), UInt8((first_unacked >> 8) & 0xff),
        nack_lo, nack_hi, stream_id,
    ]
end

"Returns `(first, last, stream_id)` or `nothing`."
function parse_heartbeat(frame::Vector{UInt8})
    (length(frame) < 13 || frame[5] != SM_HEARTBEAT) && return nothing
    first = UInt16(frame[9]) | (UInt16(frame[10]) << 8)
    last = UInt16(frame[11]) | (UInt16(frame[12]) << 8)
    (first = first, last = last, stream_id = frame[13])
end

"Returns `(first_unacked, nack_lo, nack_hi, stream_id)` or `nothing`."
function parse_acknack(frame::Vector{UInt8})
    (length(frame) < 13 || frame[5] != SM_ACKNACK) && return nothing
    first_unacked = UInt16(frame[9]) | (UInt16(frame[10]) << 8)
    (first_unacked = first_unacked, nack_lo = frame[11], nack_hi = frame[12], stream_id = frame[13])
end

# --- sender ---

@enum SubmitStatus SR_OK SR_TOO_LARGE SR_WINDOW_FULL

mutable struct Sender
    next_seq::UInt16
    in_flight::Dict{UInt16,Vector{UInt8}}
    last_heartbeat_ms::Int  # -1 = never sent
end
Sender() = Sender(UInt16(0), Dict{UInt16,Vector{UInt8}}(), -1)

in_flight_count(s::Sender) = length(s.in_flight)

"Buffers a new sample. Errors when the payload is too large or the window
(16 in-flight) is full — the caller must process ACKNACKs first."
function submit!(s::Sender, payload::Vector{UInt8})
    length(payload) > MAX_PAYLOAD && return (SR_TOO_LARGE, UInt16(0))
    length(s.in_flight) >= SENDER_WINDOW && return (SR_WINDOW_FULL, UInt16(0))
    sq = s.next_seq
    s.in_flight[sq] = payload
    s.next_seq += UInt16(1)
    (SR_OK, sq)
end

get_in_flight(s::Sender, sq::UInt16) = get(s.in_flight, sq, nothing)

"Snapshot of in-flight `(seq, payload)` pairs (safe to iterate while sending)."
in_flight_pairs(s::Sender) = collect(s.in_flight)

"`some(HEARTBEAT)` when in-flight samples exist and the period elapsed
(fires on the first call, then every `HEARTBEAT_PERIOD_MS`)."
function pending_heartbeat!(s::Sender, now_ms::Integer)
    isempty(s.in_flight) && return nothing
    due = s.last_heartbeat_ms < 0 || (now_ms - s.last_heartbeat_ms) >= HEARTBEAT_PERIOD_MS
    due || return nothing
    s.last_heartbeat_ms = now_ms
    # RFC-1982 window base (oldest unacked) + end (newest unacked); NOT numeric
    # minimum/maximum, which is wrong across a 16-bit wrap: window
    # 0xFFFE,0xFFFF,0x0000,0x0001 -> base 0xFFFE / end 0x0001, not 0x0000 /
    # 0xFFFF. Mirrors window_base / serial_max_in_flight in crates/xrce.
    mn = mx = nothing
    for k in keys(s.in_flight)
        if mn === nothing
            mn = k; mx = k
        else
            seq_lt(k, mn) && (mn = k)
            seq_gt(k, mx) && (mx = k)
        end
    end
    (first = mn, last = mx, stream_id = RELIABLE_STREAM_ID)
end

"`base = first_unacked`; everything `< base` (RFC-1982) is acknowledged and
dropped; in `[base, base+16)` a set bit means missing (keep), a clear bit
means acked (drop)."
function recv_acknack!(s::Sender, first_unacked::UInt16, nack_lo::UInt8, nack_hi::UInt8)
    base = first_unacked
    bitmap = UInt16(nack_lo) | (UInt16(nack_hi) << 8)
    to_remove = UInt16[]
    for k in keys(s.in_flight)
        diff = (UInt32(base) - UInt32(k)) & 0xffff
        if diff != 0x00000000 && diff < 0x00008000
            push!(to_remove, k)
        end
    end
    for k in to_remove
        delete!(s.in_flight, k)
    end
    for i in UInt16(0):UInt16(15)
        sq = base + i
        bit = (bitmap >> i) & UInt16(1)
        if bit == 0 && haskey(s.in_flight, sq)
            delete!(s.in_flight, sq)
        end
    end
end

# --- receiver ---

mutable struct Receiver
    expected::UInt16
    received::Dict{UInt16,Vector{UInt8}}
end
Receiver() = Receiver(UInt16(0), Dict{UInt16,Vector{UInt8}}())

expected(r::Receiver) = r.expected
out_of_order_count(r::Receiver) = length(r.received)

"Buffers an incoming sample. Returns `false` only when the receiver buffer
is full (DoS bound); duplicates are silently accepted as a no-op."
function recv_data!(r::Receiver, sq::UInt16, payload::Vector{UInt8})
    seq_lt(sq, r.expected) && return true  # already delivered → drop
    haskey(r.received, sq) && return true  # already in the buffer
    length(r.received) >= RECEIVER_BUFFER && return false
    r.received[sq] = payload
    true
end

"Delivers contiguously from `expected`, advancing `expected`."
function drain_in_order!(r::Receiver)
    out = Tuple{UInt16,Vector{UInt8}}[]
    while haskey(r.received, r.expected)
        push!(out, (r.expected, r.received[r.expected]))
        delete!(r.received, r.expected)
        r.expected += UInt16(1)
    end
    out
end

"Bitmap of the missing slots in `[expected, expected+16)`; slots beyond
`hint_last_seen` are not marked missing."
function pending_acknack(r::Receiver, hint_last_seen::Union{UInt16,Nothing})
    base = r.expected
    bitmap = UInt16(0)
    for i in UInt16(0):UInt16(15)
        sq = base + i
        if hint_last_seen !== nothing && seq_gt(sq, hint_last_seen)
            continue
        end
        if !haskey(r.received, sq)
            bitmap |= (UInt16(1) << i)
        end
    end
    (first_unacked = base, nack_lo = UInt8(bitmap & 0xff), nack_hi = UInt8((bitmap >> 8) & 0xff),
     stream_id = RELIABLE_STREAM_ID)
end

function reset!(s::Sender)
    s.next_seq = UInt16(0)
    empty!(s.in_flight)
    s.last_heartbeat_ms = -1
end

function reset!(r::Receiver)
    r.expected = UInt16(0)
    empty!(r.received)
end

end # module

File: endpoints/nim/reliable.nim
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 ZeroDDS Contributors
#
# Reliable XRCE stream (spec §8.4.10/§8.4.11) for the Nim endpoint SDK.
# Self-contained: the sender/receiver state machines + the HEARTBEAT / ACKNACK
# frame codec, byte-identical to `crates/xrce` and `endpoints/c`. Importable
# beside `zerodds.nim`; the reliable WRITE_DATA frame is the same 8-byte header
# as `zerodds.writeFrame(0x80, 0x80, seq, sample)`.
#
# Frame layout (little-endian): [0]=session [1]=stream [2..4]=seq LE
#   [4]=submessage id [5]=flags [6..8]=body-len LE [8..]=body
#   WRITE_DATA id=0x07 flags=0x03 body=sample; header seq = RFC-1982 sample seq
#   HEARTBEAT  id=0x0B flags=0x01 body(5)= first i16 LE, last i16 LE, stream
#   ACKNACK    id=0x0A flags=0x01 body(5)= first_unacked i16 LE, nack[0], nack[1], stream

import std/[tables, options]

const
  SessionNoKey* = 0x80'u8
  StreamNone* = 0x00'u8
  ReliableStreamId* = 0x80'u8
  HeartbeatPeriodMs* = 500
  SenderWindow* = 16
  ReceiverBuffer* = 64
  MaxPayload* = 65535
  SmWriteData = 0x07'u8
  SmAckNack = 0x0A'u8
  SmHeartbeat = 0x0B'u8
  FlagWrite = 0x03'u8
  FlagELe = 0x01'u8

type
  HeartbeatBody* = object
    first*, last*: uint16
    streamId*: uint8
  AckNackBody* = object
    firstUnacked*: uint16
    nackLo*, nackHi*: uint8
    streamId*: uint8
  SubmitStatus* = enum
    srOk, srTooLarge, srWindowFull

# --- RFC-1982 16-bit serial-number comparison ---

proc seqLt*(a, b: uint16): bool =
  ## `a < b` in RFC-1982 order (half-window = 32768).
  let d = (b.uint32 - a.uint32) and 0xffff
  d != 0'u32 and d < 0x8000'u32

proc seqGt*(a, b: uint16): bool =
  seqLt(b, a)

# --- frame codec ---

proc reliableWriteFrame*(seqNo: uint16, sample: seq[byte]): seq[byte] =
  ## Reliable WRITE_DATA — identical bytes to `writeFrame(0x80, 0x80, seq, sample)`.
  let n = sample.len
  result = @[SessionNoKey, ReliableStreamId,
             byte(seqNo and 0xff'u16), byte((seqNo shr 8) and 0xff'u16),
             SmWriteData, FlagWrite,
             byte(n and 0xff), byte((n shr 8) and 0xff)]
  for x in sample:
    result.add(x)

proc heartbeatFrame*(session, streamHdr: uint8, msgSeq: int, hb: HeartbeatBody): seq[byte] =
  ## byte-identical to `golden_heartbeat_le.bin` for (0x80, 0x00, 1, {1,3,0x80}).
  @[session, streamHdr, byte(msgSeq and 0xff), byte((msgSeq shr 8) and 0xff),
    SmHeartbeat, FlagELe, 5'u8, 0'u8,
    byte(hb.first and 0xff'u16), byte((hb.first shr 8) and 0xff'u16),
    byte(hb.last and 0xff'u16), byte((hb.last shr 8) and 0xff'u16),
    hb.streamId]

proc acknackFrame*(session, streamHdr: uint8, msgSeq: int, ack: AckNackBody): seq[byte] =
  ## byte-identical to `golden_acknack_le.bin` for (0x80, 0x00, 1, {1,0,0,0x80}).
  @[session, streamHdr, byte(msgSeq and 0xff), byte((msgSeq shr 8) and 0xff),
    SmAckNack, FlagELe, 5'u8, 0'u8,
    byte(ack.firstUnacked and 0xff'u16), byte((ack.firstUnacked shr 8) and 0xff'u16),
    ack.nackLo, ack.nackHi, ack.streamId]

proc parseHeartbeat*(frame: seq[byte]): Option[HeartbeatBody] =
  if frame.len < 13 or frame[4] != SmHeartbeat:
    return none(HeartbeatBody)
  some(HeartbeatBody(
    first: uint16(frame[8]) or (uint16(frame[9]) shl 8),
    last: uint16(frame[10]) or (uint16(frame[11]) shl 8),
    streamId: frame[12]))

proc parseAcknack*(frame: seq[byte]): Option[AckNackBody] =
  if frame.len < 13 or frame[4] != SmAckNack:
    return none(AckNackBody)
  some(AckNackBody(
    firstUnacked: uint16(frame[8]) or (uint16(frame[9]) shl 8),
    nackLo: frame[10], nackHi: frame[11], streamId: frame[12]))

# --- sender ---

type Sender* = object
  nextSeq: uint16
  inFlight: Table[uint16, seq[byte]]
  lastHeartbeatMs: int ## -1 = never sent

proc newSender*(): Sender =
  Sender(nextSeq: 0'u16, inFlight: initTable[uint16, seq[byte]](), lastHeartbeatMs: -1)

proc inFlightCount*(s: Sender): int = s.inFlight.len

proc submit*(s: var Sender, payload: seq[byte]): tuple[status: SubmitStatus, seq: uint16] =
  ## Buffers a new sample. Errors when the payload is too large or the window
  ## (16 in-flight) is full — the caller must process ACKNACKs first.
  if payload.len > MaxPayload:
    return (srTooLarge, 0'u16)
  if s.inFlight.len >= SenderWindow:
    return (srWindowFull, 0'u16)
  let sq = s.nextSeq
  s.inFlight[sq] = payload
  s.nextSeq = s.nextSeq + 1'u16
  (srOk, sq)

proc getInFlight*(s: Sender, sq: uint16): Option[seq[byte]] =
  if s.inFlight.hasKey(sq): some(s.inFlight[sq]) else: none(seq[byte])

iterator inFlightPairs*(s: Sender): tuple[seq: uint16, payload: seq[byte]] =
  for k, v in s.inFlight.pairs:
    yield (k, v)

proc pendingHeartbeat*(s: var Sender, nowMs: int): Option[HeartbeatBody] =
  ## `some(HEARTBEAT)` when in-flight samples exist and the period elapsed
  ## (fires on the first call, then every `HeartbeatPeriodMs`).
  if s.inFlight.len == 0:
    return none(HeartbeatBody)
  let due = s.lastHeartbeatMs < 0 or (nowMs - s.lastHeartbeatMs) >= HeartbeatPeriodMs
  if not due:
    return none(HeartbeatBody)
  s.lastHeartbeatMs = nowMs
  # RFC-1982 window base (oldest unacked) + end (newest unacked); NOT numeric
  # min/max, which is wrong across a 16-bit wrap: window
  # 0xFFFE,0xFFFF,0x0000,0x0001 -> base 0xFFFE / end 0x0001, not 0x0000 /
  # 0xFFFF. Mirrors window_base / serial_max_in_flight in crates/xrce.
  var mn = 0'u16
  var mx = 0'u16
  var seen = false
  for k in s.inFlight.keys:
    if not seen:
      mn = k; mx = k; seen = true
    else:
      if seqLt(k, mn): mn = k
      if seqGt(k, mx): mx = k
  some(HeartbeatBody(first: mn, last: mx, streamId: ReliableStreamId))

proc recvAcknack*(s: var Sender, a: AckNackBody) =
  ## `base = firstUnacked`; everything `< base` (RFC-1982) is acknowledged and
  ## dropped; in `[base, base+16)` a set bit means missing (keep), a clear bit
  ## means acked (drop).
  let base = a.firstUnacked
  let bitmap = uint16(a.nackLo) or (uint16(a.nackHi) shl 8)
  var toRemove: seq[uint16] = @[]
  for k in s.inFlight.keys:
    let diff = (base.uint32 - k.uint32) and 0xffff
    if diff != 0'u32 and diff < 0x8000'u32:
      toRemove.add(k)
  for k in toRemove:
    s.inFlight.del(k)
  for i in 0'u16 ..< 16'u16:
    let sq = base + i
    if ((bitmap shr i) and 1'u16) == 0'u16 and s.inFlight.hasKey(sq):
      s.inFlight.del(sq)

# --- receiver ---

type Receiver* = object
  expected: uint16
  received: Table[uint16, seq[byte]]

proc newReceiver*(): Receiver =
  Receiver(expected: 0'u16, received: initTable[uint16, seq[byte]]())

proc expected*(r: Receiver): uint16 = r.expected
proc outOfOrderCount*(r: Receiver): int = r.received.len

proc recvData*(r: var Receiver, sq: uint16, payload: seq[byte]): bool =
  ## Buffers an incoming sample. Returns `false` only when the receiver buffer
  ## is full (DoS bound); duplicates are silently accepted as a no-op.
  if seqLt(sq, r.expected):
    return true # already delivered → drop
  if r.received.hasKey(sq):
    return true
  if r.received.len >= ReceiverBuffer:
    return false
  r.received[sq] = payload
  true

proc drainInOrder*(r: var Receiver): seq[tuple[seq: uint16, payload: seq[byte]]] =
  ## Delivers contiguously from `expected`, advancing `expected`.
  result = @[]
  while r.received.hasKey(r.expected):
    result.add((r.expected, r.received[r.expected]))
    r.received.del(r.expected)
    r.expected = r.expected + 1'u16

proc pendingAcknack*(r: Receiver, hintLastSeen: Option[uint16]): AckNackBody =
  ## Bitmap of the missing slots in `[expected, expected+16)`; slots beyond
  ## `hintLastSeen` are not marked missing.
  let base = r.expected
  var bitmap: uint16 = 0
  for i in 0'u16 ..< 16'u16:
    let sq = base + i
    if hintLastSeen.isSome and seqGt(sq, hintLastSeen.get):
      continue
    if not r.received.hasKey(sq):
      bitmap = bitmap or (1'u16 shl i)
  AckNackBody(firstUnacked: base,
              nackLo: byte(bitmap and 0xff'u16),
              nackHi: byte((bitmap shr 8) and 0xff'u16),
              streamId: ReliableStreamId)

proc reset*(s: var Sender) =
  s.nextSeq = 0'u16
  s.inFlight.clear()
  s.lastHeartbeatMs = -1

proc reset*(r: var Receiver) =
  r.expected = 0'u16
  r.received.clear()
